procrastinate/job_processor.py

```diff
@@ -0,0 +1,225 @@
+            delete_job = {
+                DeleteJobCondition.ALWAYS: True,
+                DeleteJobCondition.NEVER: False,
+                DeleteJobCondition.SUCCESSFUL: status == Status.SUCCEEDED,
+            }[self._delete_jobs]
+            await self._job_manager.finish_job(
+                job=job, status=status, delete_job=delete_job
+            )
+
+        self.job_context = None
+        self._job_queue.task_done()
+        self.logger.debug(
+            f"Acknowledged job completion {job.call_string}",
+            extra=self._base_context.log_extra(action="finish_task", status=status),
+        )
+
+    async def run(self):
+        while True:
+            job = await self._job_queue.get()
+            async with self._job_semaphore:
+                cancelledError: asyncio.CancelledError | None = None
+                status = Status.FAILED
+                try:
+                    self.job_context = self._create_job_context(job)
+                    self.logger.debug(
+                        f"L poaded job info, about to start job {job.call_string}",
+                        extra=self.job_context.log_extra(action="loaded_job_info"),
+                    )
+                    process_job_task = asyncio.create_task(self._process_job())
+
+                    try:
+                        # the job is shielded from cancellation to enable graceful stop
+                        await asyncio.shield(process_job_task)
+                    except asyncio.CancelledError as e:
+                        cancelledError = e
+                        await process_job_task
+
+                    status = Status.SUCCEEDED
+                except TaskNotFound as exc:
+                    self.logger.exception(
+                        f"Task was not found: {exc}",
+                        extra=self._base_context.log_extra(
+                            action="task_not_found", exception=str(exc)
+                        ),
+                    )
+                except JobAborted:
+                    status = Status.ABORTED
+                except Exception:
+                    # exception is already logged by _process_job, carry on
+                    pass
+                finally:
+                    persist_job_status_task = asyncio.create_task(
+                        self._persist_job_status(job=job, status=status)
+                    )
+                    try:
+                        # prevent cancellation from stopping persistence of job status
+                        await asyncio.shield(persist_job_status_task)
+                    except asyncio.CancelledError:
+                        await persist_job_status_task
+                        raise
+
+                    async with self._fetch_job_condition:
+                        self._fetch_job_condition.notify()
+
+                    # reraise the cancelled error we caught earlier
+                    if cancelledError:
+                        raise cancelledError
```

Comment:
Never raise in Finally. It's a Python gotcha. Raise in `except` or `else`, don't worry, the exception will be raised after finally executes, but never `raise` in finally.

The reason is that it completely shadows the output (or exception) of the function.

```python
try:
    0/0
except Exception:
    print(yay)  # Should raise NameError on `yay` not being defined
finally:
    raise ValueError  # Because of this, you can never see the NameError.
```